sw/device/lib/dif/dif_i2c.c

```diff
@@ -326,173 +305,226 @@ dif_i2c_result_t dif_i2c_set_watermarks(const dif_i2c_t *i2c,
 }
 
-dif_i2c_result_t dif_i2c_irq_get(const dif_i2c_t *i2c, dif_i2c_irq_type_t type,
-                                 bool *flag_out) {
-  if (i2c == NULL || flag_out == NULL) {
+dif_i2c_result_t dif_i2c_irq_acknowledge(const dif_i2c_t *i2c,
+                                         dif_i2c_irq_t irq) {
+  if (i2c == NULL) {
     return kDifI2cBadArg;
   }
 
-  uint32_t index;
-  dif_i2c_result_t err = irq_bit_for_type(type, &index);
-  if (err != kDifI2cOk) {
-    return err;
+  bitfield_bit32_index_t index;
+  if (!irq_index(irq, &index)) {
+    return kDifI2cBadArg;
   }
 
-  *flag_out =
-      mmio_region_get_bit32(i2c->base_addr, I2C_INTR_STATE_REG_OFFSET, index);
+  uint32_t reg = bitfield_bit32_write(0, index, true);
+  mmio_region_write32(i2c->params.base_addr, I2C_INTR_STATE_REG_OFFSET, reg);
 
   return kDifI2cOk;
 }
 
-dif_i2c_result_t dif_i2c_irq_clear(const dif_i2c_t *i2c,
-                                   dif_i2c_irq_type_t type) {
+dif_i2c_result_t dif_i2c_irq_get_enabled(const dif_i2c_t *i2c,
+                                         dif_i2c_irq_t irq,
+                                         dif_i2c_toggle_t *state) {
   if (i2c == NULL) {
     return kDifI2cBadArg;
   }
 
-  uint32_t index;
-  dif_i2c_result_t err = irq_bit_for_type(type, &index);
-  if (err != kDifI2cOk) {
-    return err;
+  bitfield_bit32_index_t index;
+  if (!irq_index(irq, &index)) {
+    return kDifI2cBadArg;
   }
 
-  // IRQ state registers are write-one-clear.
-  mmio_region_write_only_set_bit32(i2c->base_addr, I2C_INTR_STATE_REG_OFFSET,
-                                   index);
+  uint32_t reg =
+      mmio_region_read32(i2c->params.base_addr, I2C_INTR_ENABLE_REG_OFFSET);
+  bool is_enabled = bitfield_bit32_read(reg, index);
+  *state = is_enabled ? kDifI2cToggleEnabled : kDifI2cToggleDisabled;
 
   return kDifI2cOk;
 }
 
 dif_i2c_result_t dif_i2c_irq_set_enabled(const dif_i2c_t *i2c,
-                                         dif_i2c_irq_type_t type,
-                                         dif_i2c_enable_t state) {
+                                         dif_i2c_irq_t irq,
+                                         dif_i2c_toggle_t state) {
   if (i2c == NULL) {
     return kDifI2cBadArg;
   }
 
-  uint32_t index;
-  dif_i2c_result_t err = irq_bit_for_type(type, &index);
-  if (err != kDifI2cOk) {
-    return err;
+  bitfield_bit32_index_t index;
+  if (!irq_index(irq, &index)) {
+    return kDifI2cBadArg;
   }
 
+  bool flag;
   switch (state) {
-    case kDifI2cEnabled:
-      mmio_region_nonatomic_set_bit32(i2c->base_addr,
-                                      I2C_INTR_ENABLE_REG_OFFSET, index);
+    case kDifI2cToggleEnabled:
+      flag = true;
       break;
-    case kDifI2cDisabled:
-      mmio_region_nonatomic_clear_bit32(i2c->base_addr,
-                                        I2C_INTR_ENABLE_REG_OFFSET, index);
+    case kDifI2cToggleDisabled:
+      flag = false;
       break;
     default:
       return kDifI2cBadArg;
   }
```

Comment:
You may want to refactor this to a function e.g. `toggle_to_bool` (4 places).

Reply:
Not sure I agree. I would still have to check the result of such a function, and doing the whole ceremony of an out-param feels unnecessary.